tests/utils_tests/test_safestring.py

```diff
@@ -98,3 +118,44 @@ def html_str():
 
         lazy_str = lazy(html_str, str)()
         self.assertEqual(mark_safe(lazy_str), html_str())
+
+
+class SafeStringSlotsTest(SimpleTestCase):
+    def test_default_safe_string_additional_attrs(self):
+        s = SafeString("a&b")
+        with self.assertRaisesMessage(
+            AttributeError, "object has no attribute 'dynamic_attr'"
+        ):
+            s.dynamic_attr = True
+
+    def test_custom_safe_data_has_dict(self):
+        """
+        Adding __slots__ to SafeData to reduce memory-usage of SafeString
+        doesn't prevent expected subclass behaviour.
+        """
+        s = CustomSafeData(True)
+        s.dynamic_attr = True
+        self.assertTrue(hasattr(s, "__dict__"))
+        self.assertIs(s.cls_attr, True)
+        self.assertIs(s.instance_attr, True)
+        self.assertIs(s.instance_cached_attr, True)
+        self.assertIs(s.dynamic_attr, True)
+        self.assertNotIn("cls_attr", s.__dict__)
+        self.assertIn("instance_attr", s.__dict__)
+        self.assertIn("instance_cached_attr", s.__dict__)
+        self.assertIn("dynamic_attr", s.__dict__)
+
+    def test_custom_safe_string_has_dict(self):
+        """
+        Adding __slots__ to SafeData & SafeString to reduce memory-usage
+        doesn't prevent expected subclass behaviour.
+        """
+        s = CustomSafeString("a&b")
+        s.dynamic_attr = True
+        self.assertTrue(hasattr(s, "__dict__"))
+        self.assertIs(s.cls_attr, True)
+        self.assertIs(s.instance_cached_attr, True)
+        self.assertIs(s.dynamic_attr, True)
+        self.assertNotIn("cls_attr", s.__dict__)
+        self.assertIn("instance_cached_attr", s.__dict__)
+        self.assertIn("dynamic_attr", s.__dict__)
```

Comment:
Do we need these tests? As far as I'm aware we're testing Python's behavior here, not Django :thinking: I'd leave only `test_default_safe_string_additional_attrs()` and add the same for `SafeData`.

Reply:
Nah, we don't _need_ these tests, they're absolutely just exercising "this is how [c]python works" ... but it was easier for me to add them to demonstrate the behaviour up-front, in case one cannot recall the intricacies of `__slots__` in subclasses. Show don't tell, I guess :)

I'll happily remove them and drop down to just the `AttributeError` test(s).

Reply:
I agree to removing them, no need to test python behaviour